docs/proposals/emptydir-pv.md

```diff
@@ -0,0 +1,151 @@
+* Every pod referencing this PV must land on the same node (this concept has been referred to as "data gravity").
+* Once this PV is bound to a node, it cannot get unbound till a human intervenes or the node is deleted
+* The PV must require low user interaction. It should be provisionable dynamically.
+* The directory backing the PV must get deleted when the PV is deleted.
+
+## Non goals
+* A local storage abstraction that survives extended node outages
+* Snapshotting or simply copying the local volume to another node
+
+## A new PV storage source
+
+As described above, the storage source for the PV must be a local volume. While we already allow hostPath as a PersistentVolumeSource, it is insufficient because:
+* user needs to specify the exact path, thus requires knowledge of the filesystem
+* hard to dynamically provision a hostpath
+* hostpath volumes are not cleaned up
+
+The `emptydir` abstraction solves these problems:
+* it is allocated in `/var/lib/kubelet/resource/uid/`, a location that is predictable but unknow till `resource` creation time
+* The node takes care of allocation, a dynamic provisioner just needs to create the PV
+* it is cleaned up when the pod dies (or in this case when the PV is deleted)
+
+We can create a new [PersistetVolumeSource](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L240) for EmptyDir. Users will request storage by creating PersistetVolumeClaims. If the PVC isn't bound to a PV, one will be created dynamically. The user can't control which node the storage for this PV comes from, or where in the node's filesystem it is allocated. Users can also hand create PVs bound to specific nodes (TODO: elaborate on QoS and what happens if there's no space on the node).
+
+## Scheduling constraints for data gravity
+
+The scheduler will schedule the pod paying attention to any referenced PVs. If an empty dir PV attached to an existing node is found, that trumps all scheduler predicates. Otherwise the scheduler makes a scheduling decision taking into account pod cpu/memory requests (and in the future, requested storage), attaches it to a node, and all subsequent scheduling decisions for any pod referencing the same PV evaluate to the same node. The scheduler "Attaches" a pv to a node by writing an annotation (TBD what actually writes this annotation, scheduler or kubelet, scheduler will at least bind the pod to the node). All pods referencing the PV that enter the "Running" state on the give node will be able to write to the volume. If the node enters the `NotReady` state, all pods referencing the PV attached to the node will remain in `Pending` till such time when an admin either removes the annotation, or rewrites it to point to another node. The directory allocated to the emptydir volume will be exported through the status field of the PV, so admins can yank the disk from a NotReady node and rsync data to the same directory on another node before rewriting the annotation.
```

Comment:
We could also teach the volume controller to pick the host the first pod lands on as "the host" and mark other pods as Failed. So if i understand your idea correctly:   

1. Scheduler puts pod on some host
2. Kubelet waits for volume controller to add labels `pv=pv-uuid` to it's own node object ("attach")
3. Kubelet starts pod (we need to figure out some place to put creation of the pv dir, since attach is used by the volume controller. Lets ignore this for now.)
4. Either all the kubelets watch all nodes and notice that they have a pod with a local volume bound to another node, or the volume controller itself understands this, and the other pods are marked Failed
5. Node NotReady, all pods get deleted
6. pods rescheduled on another node
7. Volume controller copies data from dead node onto new node ("detach")
8. Volume controller applies label (second "attach")

Vs 

1. Scheduler puts pod on some host, sets hostname field
2. Kubelet checks hostname == nodename, sets Failed otherwise
3. Kubelet attaches by creating directory for pv, and bind mounts directory with AccessMode
4. Kubelet starts pod
5. Node NotReady, all pods get deleted
6. pods DONT rescheduled on another node, scheduler can't find a node for the pods
7. Volume controller copies data from dead node onto new node (unnamed step?)
8. Volume controller resets hostname annotation
9. Scheduler now finds new node for pods

Correct?  
On the one hand we already know and understand label based scheduling, and on the other, using labels makes exclusive 1:1 mappings slightly harder to deal with.